packages/eslint-plugin/src/rules/no-reference-import.ts

```diff
@@ -0,0 +1,73 @@
+import * as util from '../util';
+import { AST_NODE_TYPES } from '@typescript-eslint/typescript-estree';
+import {
+  Literal,
+  Node,
+  TSExternalModuleReference,
+} from '@typescript-eslint/typescript-estree/dist/ts-estree/ts-estree';
+
+export default util.createRule({
+  name: 'no-reference-import',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Disallow simultaneous use of `/// <reference type="" />` comments and ES6 style imports for the same module',
+      category: 'Best Practices',
+      recommended: false,
+    },
+    schema: [],
+    messages: {
+      noReferenceImport: 'Do not reference {{module}} if importing it anyway.',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    let programNode: Node;
+
+    function hasMatchingReference(source: Literal) {
+      const referenceRegExp = /^\/\s*<reference\s*types="(.*)"/;
+      const sourceCode = context.getSourceCode();
+      const commentsBefore = sourceCode.getCommentsBefore(programNode);
+
+      commentsBefore.forEach(comment => {
+        if (comment.type !== 'Line') {
+          return;
+        }
+        const referenceResult = referenceRegExp.exec(comment.value);
```

Comment:
It seems like you're running this every time you see an import node.
Could you run this once and extract the list of referenced packages from the references, then look up in that list on each import node?

Reply:
Moved the collecting references bit to the program node.